File: 0290-word-pattern/0290-word-pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        stringstream ss(s);
        vector<string>arr;
        while(ss>>s){
            arr.push_back(s);
        }
        
        if(arr.size()!=pattern.size())
            return false;
        
        unordered_map<char,string>cs;
        unordered_map<string,char>sc;
        
        for(int i=0;i<arr.size();i++){
            if(cs.find(pattern[i])!=cs.end() && cs[pattern[i]]!=arr[i])
                return false;
            if(sc.find(arr[i])!=sc.end() && sc[arr[i]]!=pattern[i])
                return false;
            cs[pattern[i]]=arr[i];
            sc[arr[i]]=pattern[i];
        }
        
        return true;
    }
};

File: 0290-word-pattern/0290-word-pattern.java
class Solution {
    
    
    public boolean wordPattern(String pattern, String s) {
        Map<Character,String>m1=new HashMap<>();
        Map<String,Character>m2=new HashMap<>();
        
        String arr[]=s.split(" ");
        
        int len=pattern.length();
        
        if(arr.length!=len){
            return false;
        }
        
        for(int i=0;i<len;i++){
            
            if(m1.get(pattern.charAt(i))==null){
                m1.put(pattern.charAt(i),arr[i]);
            }
            if(m2.get(arr[i])==null){
                m2.put(arr[i],pattern.charAt(i));
            }
            
            if(!m1.get(pattern.charAt(i)).equals(arr[i]) || !m2.get(arr[i]).equals(pattern.charAt(i))){
                return false;
            }
        }
        
        return true;
    }
}